drivers: sensor: ti: Add INA228

wilkinsw opened the pull request:
Add the Texas Instruments INA228 Power/Energy/Charge monitor. This chip is similar to the INA237, but has a higher resolution ADC and energy accumulator.

Datasheet: https://www.ti.com/lit/ds/symlink/ina228.pdf

Edit: Ok, this spiraled a bit out of control and is now a reorganization/refactor of all the INA2XX sensors since they all pretty much do the same thing in the same way.

pyhys committed:
drivers: sensor: ti: ina2xx: Add INA228 test

Add test for Texas Instruments INA228 current sensor via I2C.

Signed-off-by: Jonas Berg <jonas.s.t.berg@gmail.com>
Signed-off-by: Wilkins White <ww@novadynamics.com>

wilkinsw commented:
Ah this was PR'd two days ago too:
https://github.com/zephyrproject-rtos/zephyr/pull/92365#issue-3186334058

wilkinsw commented:
The differences are in get, fetch, and the calibration register. 

The simplest approach would be to have two sensor API tables that use the common init and attr functions but the specific get/fetch.

It wouldn't be hard to do, I can look at it next week unless @pyhys gets to it first.

pyhys commented:
Great! My main interest is to have mainline support for INA228, so I like your approach of expanding the existing driver. 
I think it would be great if there could be official SENSOR_CHAN_ENERGY and SENSOR_CHAN_CHARGE channels. Please see my PR (I do not know if it is the right way to do it).

Please reuse my test cases for emulated I2C communication.

wilkinsw commented:
My boards with the INA228 on them arrive next week, so I expect I'll be implementing the suggested changes soon.

wilkinsw committed:
drivers: sensor: ti: Rename ina23x to ina2xx

This commit renames the ina23x folder and common files to ina2xx.
The more generic name will support, e.g., the INA237 and INA228
series chips.

Signed-off-by: Wilkins White <ww@novadynamics.com>

wilkinsw commented:
@ubieda are these changes more in line with what you were looking for?

wilkinsw committed:
drivers: sensor: ti: ina2xx: Convert INA237

Convert the INA237 driver to INA2XX

Signed-off-by: Wilkins White <ww@novadynamics.com>

wilkinsw committed:
drivers: sensor: ti: ina2xx: Convert INA230

Convert the INA230 driver to INA2XX

Signed-off-by: Wilkins White <ww@novadynamics.com>

wilkinsw commented:
Ok, wow, this got a bit out of hand but all the sensors are now under the INA2XX namespace and inherit `ti,ina2xx-common.yaml` as well as using some common initialization/configuration now.

I don't have most of these to test, so I tried not to touch any of the functional parts.  There's also definitely more that could be done, but it's a start.

Going to open this back up while I wait for the CI to make me feel bad about myself.

wilkinsw commented:
TODO: go through https://github.com/zephyrproject-rtos/zephyr/pull/92365 and make sure I have all the same stuff

bperseghetti commented:
> I don't have most of these to test, so I tried not to touch any of the functional parts.  There's also definitely more that could be done, but it's a start.

I have a 226 and a 228 that I can test with.

wilkinsw commented:
I'm going to separate the re-path (INA219, INA226, etc) from the actual file changes to make this parse a bit easier

wilkinsw commented:
> > I don't have most of these to test, so I tried not to touch any of the functional parts.  There's also definitely more that could be done, but it's a start.
> 
> I have a 226 and a 228 that I can test with.

My 228 is pretty happy:

49.3V, at ~25 mA for half a second is about 0.6 joules

```
[00:03:53.577,000] <inf> monitor: bus=49.3V, sys=0.0V, mot=0.0V, i=0.027A, E=0.652J, soc=0%
[00:04:03.688,000] <inf> monitor: bus=49.3V, sys=0.0V, mot=0.0V, i=0.022A, E=0.609J, soc=0%
```

Here's the dtsi I'm using:
```
	ina228: ina228@40 {
		compatible = "ti,ina228";
		reg = <0x40>;
		status = "okay";
		current-lsb-microamps = <850>;
		rshunt-micro-ohms = <2000>;
	};
```

and

```
#include <zephyr/drivers/sensor/ina2xx.h>
if(sensor_channel_get(ina2xx, SENSOR_CHAN_INA2XX_ENERGY, &data) == 0) {
    ...
}
```

ubieda reviewed (commented):
Skimming through the patches, at a glance looks good. Let's address CI and SonarQube issues. Then we can kick-off the review process!

ubieda commented:
Perhaps ping the contributors for previous INA drivers to have them test this PR on their target?

wilkinsw committed:
drivers: sensor: ti: ina2xx: Convert INA226

Convert the INA226 driver to INA2XX

Signed-off-by: Wilkins White <ww@novadynamics.com>

wilkinsw committed:
drivers: sensor: ti: ina2xx: Add INA228

Add the Texas Instruments INA228 Power/Energy/Charge
monitor. This chip is similar to the INA237, but has
a more precise ADC and added energy/charge channels.

Signed-off-by: Wilkins White <ww@novadynamics.com>

wilkinsw committed:
drivers: sensor: ti: ina2xx: Add common functions

This commit adds a common framework to reduce repeated
code in the INA2XX family.

Signed-off-by: Wilkins White <ww@novadynamics.com>

ubieda reviewed (approved):
Thanks for the hard work, LGTM.

Unfortunately I don't have hardware to test these changes. If anyone can/wants to validate this PR and drop a comment it'd be greatly appreciated.

bperseghetti reviewed (approved):
Thanks for this great add!
Works on my INA226.
